sql/migrate/dir.go

```diff
@@ -555,3 +556,73 @@ func (m *memFile) Mode() fs.FileMode          { return 0 }
 func (m *memFile) ModTime() time.Time         { return time.Time{} }
 func (m *memFile) IsDir() bool                { return false }
 func (m *memFile) Sys() interface{}           { return nil }
+
+// ArchiveDir returns a tar archive of the given directory.
+func ArchiveDir(dir Dir) ([]byte, error) {
+	var buf bytes.Buffer
+	tw := tar.NewWriter(&buf)
+	defer tw.Close()
+
+	sumF, err := dir.Open(HashFileName)
+	if err != nil && !errors.Is(err, fs.ErrNotExist) {
+		return nil, err
+	}
+	if sumF != nil {
+		sumB, err := io.ReadAll(sumF)
+		if err != nil {
+			return nil, err
+		}
+		if err := append2Tar(tw, HashFileName, sumB); err != nil {
+			return nil, err
+		}
+	}
+	files, err := dir.Files()
+	if err != nil {
+		return nil, err
+	}
+	for _, f := range files {
+		if err := append2Tar(tw, f.Name(), f.Bytes()); err != nil {
+			return nil, err
+		}
+	}
+	return buf.Bytes(), nil
+}
+
+// UnarchiveDir extracts the tar archive into the given directory.
+func UnarchiveDir(arc []byte) (Dir, error) {
+	var (
+		md = &MemDir{}
+		tr = tar.NewReader(bytes.NewReader(arc))
+	)
+	for {
+		h, err := tr.Next()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return nil, err
+		}
+		data, err := io.ReadAll(tr)
+		if err != nil {
+			return nil, err
+		}
+		if err := md.WriteFile(h.Name, data); err != nil {
+			return nil, err
+		}
+	}
+	return md, nil
```

Comment:
Add migrate.Validate here.

Reply:
I think it is fine to not validate. Esepcially if the dir archived was not valid. It should return exactly what was compressed in the first place.